### DISCUSSION TITLE
Improving SSE handling for better extensibility

### DISCUSSION THREAD

**Discussion by mirshko** [honojs/hono#3472] [Q&A] [answered]
I'm working on a small Deno server that I want to essentially lift some logic out our app and into a Server-Sent Event client (we only need one-way, so Websockets feel overkill for this)

The app uses https://viem.sh/docs/actions/public/watchBlocks#usage and my thinking is to have this `watchBlocks` function be inside of a Hono route. 

Something like

```ts
app.get("/sse", async (c) => {
  const target = c.sendEvents();

  const unwatch = publicClient.watchBlocks( 
    { onBlock: block => target.dispatchMessage(block); }
  )
  
  target.addEventListener("close", (evt) => {
    unwatch()
  });
  
  // ...
})
```

This takes the API from Oak (https://oakserver.github.io/oak/sse.html) on how they handle server sent events which seems to be quite flexible compared to the streamSSE helper from Hono (https://hono.dev/docs/helpers/streaming#streamsse)

There doesn't seem to be a way to have these long-running watch functions that output data in intervals due to the streamSSE requiring the stream.sleep() function otherwise it closes the connection (from what I understand reading other issues about SSE in Hono's context)

Would love to see what y'all have any idea on how to solve this.

**Comment by maou-shonen** [comment]
I think you're talking about two different things.
SSE is an HTTP protocol used for one-way communication between the server and the browser. Event Emitter, on the other hand, is a solution typically used within the same runtime for event-driven communication.
These two things can also be combined and used together.

Additionally, hono has an [Event Emitter middleware for Hono](https://github.com/honojs/middleware/tree/main/packages/event-emitter)

**Reply by mirshko** [reply]
Hey thanks for this, yeh I think you're right here. However, as you said, it looks like this middleware is only useful for events in the same application? 

I'm looking for a way to create an endpoint in Hono thats consumed in `EventSource` on the client. Oak calls it Server-Sent-Events in their case: https://oakserver.github.io/oak/sse.html.

From what I can tell that middleware won't work for this, and neither will the streamSSE for my use-case.

**Reply by maou-shonen** [reply]
You can wrap the EventEmitter into a stream in some way, then send it using hono/stream, and set the correct SSE headers and content format.

**Reply by maou-shonen** [reply, marked as answer]
The hono official could indeed support this. Maybe you could open a feature request?

**Reply by mirshko** [reply]
Done! https://github.com/honojs/hono/issues/3527
